File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = []
        i = 0
        while i < len(s):
            if (s[i] == "}"):
                b = stack.pop()
                if b != "{":
                    return False
            elif (s[i] == "]"):
                b = stack.pop()
                if b != "[":
                    return False
            elif (s[i] == ")"):
                b = stack.pop()
                if b != "(":
                    return False
            else:
                stack.append(s[i])
            i += 1
        return True if len(stack) == 0 else False

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        // My Implementation
        Stack<Character> stack = new Stack<>();
        // stack.push()/.pop/.peek/.empty/.search()
        int i = 0;
        while (i < s.length()) {
            if (s.charAt(i) == '(' || s.charAt(i) == '{' || s.charAt(i) == '[')
                stack.push(s.charAt(i++));
            else {
                if (stack.empty())
                    return false;

                char x = stack.pop();
                
                if (s.charAt(i) == ')' && x == '(') {
                    i++;
                    continue;
                }
                else if (s.charAt(i) == '}' && x == '{') {
                    i++;
                    continue;
                }
                else if (s.charAt(i) == ']' && x == '[') {
                    i++;
                    continue;
                }
                else
                    return false;
            }
        }
        return stack.empty();

        // ChatGPT
        // Stack<Character> stack = new Stack<>();
        
        // for (char c : s.toCharArray()) {
        //     if (c == '(' || c == '{' || c == '[') {
        //         stack.push(c);
        //     } else if (c == ')' && !stack.isEmpty() && stack.peek() == '(') {
        //         stack.pop();
        //     } else if (c == '}' && !stack.isEmpty() && stack.peek() == '{') {
        //         stack.pop();
        //     } else if (c == ']' && !stack.isEmpty() && stack.peek() == '[') {
        //         stack.pop();
        //     } else {
        //         return false;
        //     }
        // }
        
        // return stack.isEmpty();
    }
}